casper/src/main/scala/coop/rchain/casper/SynchronyConstraintChecker.scala

```diff
@@ -35,40 +37,59 @@ final class SynchronyConstraintChecker[F[_]: Sync: BlockStore: Log](
-          lastProposedTuplespace = ProtoUtil.postStateHash(lastProposedBlock)
-          bonds                  <- runtimeManager.computeBonds(lastProposedTuplespace)
-          activeValidators       <- runtimeManager.getActiveValidators(lastProposedTuplespace)
-          validatorWeightMap = bonds
-            .filter(b => activeValidators.contains(b.validator))
-            .map(b => b.validator -> b.stake)
-            .toMap
-          sendersWeight = seenSenders.toList.flatMap(s => validatorWeightMap.get(s)).sum
-          // This method can be called on readonly node or not active validator.
-          // So map validator -> stake might not have key associated with the node,
-          // that's why we need `getOrElse`
-          otherValidatorsWeight = validatorWeightMap.values.sum - validatorWeightMap
-            .getOrElse(validator, 0L)
-          // If there is no other active validators, do not put any constraint (value = 1)
-          synchronyConstraintValue = if (otherValidatorsWeight == 0) 1
-          else
-            sendersWeight.toDouble / otherValidatorsWeight
-          _ <- Log[F].info(
-                s"Seen ${seenSenders.size} senders with weight $sendersWeight out of total $otherValidatorsWeight " +
-                  s"(${synchronyConstraintValue} out of $synchronyConstraintThreshold needed)"
-              )
-        } yield synchronyConstraintValue >= synchronyConstraintThreshold
+          lastProposedBlockMeta <- dag.lookupUnsafe(lastProposedBlockHash)
+
+          checkConstraint = for {
+            // Estimate parent blocks for proposing block
+            tipHashes             <- Estimator[F].tips(dag, approvedBlock)
+            estimatedParentBlocks <- EstimatorHelper.chooseNonConflicting(tipHashes, dag)
+
+            // Get main parent of the main parent block
+            mainParentOpt  = estimatedParentBlocks.headOption
+            mainParentMeta <- mainParentOpt.liftTo[F](new Exception(s"Parent blocks not found.}"))
+
+            // Loading the whole block is only needed to get post-state hash
+            mainParentBlock     <- BlockStore[F].getUnsafe(mainParentMeta.blockHash)
+            mainParentStateHash = ProtoUtil.postStateHash(mainParentBlock)
+
+            // Get bonds map from PoS
+            // NOTE: It would be useful to have active validators cached in the block in the same way as bonds.
+            activeValidators <- runtimeManager.getActiveValidators(mainParentStateHash)
+
+            // Validators weight map filtered by active validators only.
+            validatorWeightMap = lastProposedBlockMeta.weightMap.filter {
+              case (validator, _) => activeValidators.contains(validator)
+            }
+            // Guaranteed to be present since last proposed block was present
+            seenSenders   <- calculateSeenSendersSince(lastProposedBlockMeta, dag)
+            sendersWeight = seenSenders.toList.flatMap(validatorWeightMap.get).sum
```

Comment:
@nzpr I'm not sure is it possible here that `validatorWeightMap` doesn't have validator calculated as seen sender?

Reply:
@nzpr nvm, this is already handled.

Reply:
Fix in PR #3274.